File: samples/python/mode7.py
"""
Tilengine python example:
    Raster based affine transform to simulate perspective, SNES Mode-7 style
"""

from tilengine import *
from math import sin, cos, radians

# constants
WIDTH = 400
HEIGHT = 240

# module variables
pos_x = -136
pos_y = 336
speed = 0
angle = 0


class affine:
	angle = 0
	sx = 1
	sy = 1
	dx = WIDTH / 2
	dy = HEIGHT


# linear interpolation
def lerp(pos_x, x0, x1, fx0, fx1):
	return fx0 + (fx1 - fx0) * (pos_x - x0) / (x1 - x0)


def raster_effect(line):
	if line == 24:
		background.setup(map_track)
		background.set_position(pos_x, pos_y)
		foreground.disable()

	if line > 24:
		scale = lerp(line, 24, HEIGHT, 0.2, 5.0)
		affine.sx = scale
		affine.sy = scale
		background.set_transform(affine.angle, affine.dx, affine.dy, affine.sx, affine.sy)


# setup engine
engine = Engine.create(WIDTH, HEIGHT, 2, 0, 0)
engine.set_background_color(Color(0, 0, 0))
foreground = engine.layers[0]
background = engine.layers[1]

# load resources
engine.set_load_path("../assets/smk")
map_horizon = Tilemap.fromfile("track1_bg.tmx")
map_track = Tilemap.fromfile("track1.tmx")

# set raster callback
engine.set_raster_callback(raster_effect)

# main loop
window = Window.create()
while window.process():
	foreground.setup(map_horizon)
	background.setup(map_horizon)
	foreground.set_position(lerp(angle * 2, 0, 360, 0, 256), 24)
	background.set_position(lerp(angle, 0, 360, 0, 256), 0)
	background.reset_mode()

	# input: turn left/right
	if window.get_input(Input.LEFT):
		angle -= 2
	elif window.get_input(Input.RIGHT):
		angle += 2

	# input: accelerate forward
	if window.get_input(Input.UP):
		speed = min(speed + 0.2, 2)
	elif speed > 0:
		speed = max(speed - 0.1, 0)

	# input: accelerate backwards
	if window.get_input(Input.DOWN):
		speed = max(speed - 0.2, -2)
	elif speed < 0:
		speed = min(speed + 0.1, 0)

	# displacement
	if speed != 0:
		angle = (angle + 360) % 360
		rad = radians(angle)
		pos_x += (sin(rad) * speed)
		pos_y -= (cos(rad) * speed)

	affine.angle = angle

File: samples/Mode7.c
/******************************************************************************
*
* Tilengine sample
* 2015 Marc Palacios
* http://www.tilengine.org
*
* This example shows a classic Mode 7 perspective projection plane like the 
* one seen in SNES games like Super Mario Kart. It uses a single transformed
* layer with a raster effect for setting the scaling factor on each line
*
******************************************************************************/

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "Tilengine.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#define WIDTH	400
#define HEIGHT	240

/* linear interploation */
#define lerp(x, x0,x1, fx0,fx1) \
	(fx0) + ((fx1) - (fx0))*((x) - (x0))/((x1) - (x0))
	
/* layers */
enum
{
	LAYER_FOREGROUND,
	LAYER_BACKGROUND,
	MAX_LAYER
};

enum
{
	MAP_HORIZON,
	MAP_TRACK,
	MAX_MAP
};

static TLN_Tilemap road, horizon;

static void raster_callback (int line);
static void main_loop(uint32_t frame);

int angle = 0;			/* angle */
float x = -136;			/* horizontal position */
float y = 336;			/* vertical position */
float speed = 0;		/* speed */
float accel = 0.2f;		/* acceleration */
float maxspeed = 2.0f;	/* max speed */

/* entry point */
int main (int argc, char* argv[])
{
	/* setup engine */
	TLN_Init (WIDTH,HEIGHT, MAX_LAYER, 0, 0);
	TLN_SetRasterCallback (raster_callback);
	TLN_SetBGColor (0,0,0);

	/* load resources*/
	TLN_SetLoadPath ("assets/smk");
	road = TLN_LoadTilemap ("track1.tmx", NULL);
	horizon = TLN_LoadTilemap ("track1_bg.tmx", NULL);

	/* create window & main loop, block until window closes */
	TLN_CreateWindow(NULL, CWF_FULLSCREEN);
	TLN_SetMainTask(main_loop);

	/* deinit */
	TLN_DeleteTilemap (road);
	TLN_DeleteTilemap (horizon);
	TLN_DeleteWindow ();
	TLN_Deinit ();
	return 0;
}

/* main loop delegate, called every frame */
static void main_loop(uint32_t frame)
{
	/* draws top horizon based on viewing angle */
	TLN_SetLayerTilemap (LAYER_FOREGROUND, horizon);
	TLN_SetLayerTilemap (LAYER_BACKGROUND, horizon);
	TLN_SetLayerPosition (LAYER_FOREGROUND, lerp(angle*2, 0,360, 0,256), 24);
	TLN_SetLayerPosition (LAYER_BACKGROUND, lerp(angle, 0,360, 0,256), 0);
	TLN_ResetLayerMode (LAYER_BACKGROUND);

	/* rotate left/right */		
	if (TLN_GetInput (INPUT_LEFT))
		angle -= 2;
	else if (TLN_GetInput (INPUT_RIGHT))
		angle += 2;
	
	/* move forward */
	if (TLN_GetInput (INPUT_UP))
	{
		speed += accel;
		if (speed > maxspeed)
			speed = maxspeed;
	}
	else if (speed >= accel)
		speed -= accel;
	
	/* move backwards */
	if (TLN_GetInput (INPUT_DOWN))
	{
		speed -= accel;
		if (speed < -maxspeed)
			speed = -maxspeed;
	}
	else if (speed <= -accel)
		speed += accel;
	
	if (fabs(speed) < 0.2f)
		speed = 0;

	/* normalize angle in range 0 - 360 */
	angle = (angle + 360) % 360;	

	/* move according to speed and direction */
	if (speed != 0)
	{
		float radians = (angle * M_PI) / 180.0f;
		x += sin(radians) * speed;
		y -= cos(radians) * speed;
	}
}

/* raster callback (virtual HBLANK) */
static void raster_callback (int line)
{
	if (line == 24)
	{
		TLN_SetLayerTilemap (LAYER_BACKGROUND, road);
		TLN_SetLayerPosition (LAYER_BACKGROUND, (int)x, (int)y);
		TLN_DisableLayer (LAYER_FOREGROUND);
	}

	if (line >= 24)
	{
		float scale = lerp (line, 24,HEIGHT, 0.2f,5.0f);
		TLN_SetLayerTransform (LAYER_BACKGROUND, (float)angle, WIDTH/2, HEIGHT, scale, scale);
	}
}
